cogs/rolehelper.py

```diff
@@ -44,6 +45,80 @@ async def _set_master(self, ctx) -> None:
+    async def update_years(self, ctx, *args) -> None:
+        """
+        Updates the year roles of everyone in the server (that joined before last August)
+        Optional arguments: "force" to also upgrade people that joined after last August, "downgrade" to downgrade
+        """
+        async with ctx.channel.typing():
+            largs = [s.lower() for s in args]
+            last_august = self._get_last_date(8, 1)
+            count = 0
+            guild = ctx.channel.guild
+            too_old = []
+
+            for member in guild.members:
+                if "force" not in largs:
+                    # If the member joined after last august
+                    if member.joined_at > last_august:
+                        continue
+
+                try:
+                    # Increase or decrease the roles
+                    if "downgrade" in largs:
+                        await self._change_year(member, -1)
+                    else:
+                        await self._change_year(member, 1)
+                except MemberIsTooOld:
+                    too_old.append(member.display_name)
+                    continue
+                except MemberHasNoYearRole:
+                    continue
+
+                count += 1
+
+        await ctx.send("Changed the ranks of "+str(count) +
+                       " members (this server has a total size of " + str(guild.member_count) + ")")
+        if len(too_old) > 0:
+            await ctx.send("Please manually change the roles of " + ", ".join(too_old))
+
+    @staticmethod
+    async def _change_year(member: discord.Member, amount: int = 1):
+        # Get the current year of the user
+        regex = re.compile("^Year [0-9]+$")
+        try:
+            current_role = [role for role in member.roles if regex.match(role.name)][0]
+        except IndexError:  # If index 0 is out of range; the member does not have a year role
+            raise MemberHasNoYearRole
+        current_year = int(re.findall("[0-9]+", current_role.name)[0])
+
+        # Get the new role of the user
+        new_name = "Year " + str(current_year + amount)
+        try:
+            new_role = [role for role in member.guild.roles if role.name == new_name][0]
+        except IndexError:  # If index 0 is out of range; a role with that year does not exist yet
+            raise MemberIsTooOld
+
+        # Remove old role and assign new role
+        await member.remove_roles(current_role)
+        await member.add_roles(new_role)
+
+
+    @staticmethod
+    def _get_last_date(month: int, day: int) -> datetime:
+        """
+        Gets the last occurrence of a date (month and date)
+        :return: a datetime object on the last occurrence of the provided date
+        """
+        today = datetime.now()
+        if today.month >= month and today.day >= day:
+            return datetime(today.year, month, day)
+        else:
+            return datetime(today.year - 1, month, day)
+
```

Comment:
I'm not approving this one ;) remove the update_year stuff. peopel can just do that themselves.

Reply:
The point of a bot is so people have to do less themselves smh

Reply:
oh fuck i see what you mean, i forked it wrong